### PR TITLE
ARROW-10102: [C++] Refactor BasicDecimal128 Multiplication to use unsigned helper

### PR DESCRIPTION
**Pull request by Luminarys**

### CODE REVIEW

File: cpp/src/arrow/util/basic_decimal.cc
Comment (MingyuZhong) [inline cpp/src/arrow/util/basic_decimal.cc:256]: Use kInt64Mask?

File: cpp/src/arrow/util/basic_decimal.cc
Comment (MingyuZhong) [inline cpp/src/arrow/util/basic_decimal.cc:283]: Shouldn't we use "<<" rather than ">>"? Same below. Also, we can replace "+" with "|".

**Review by MingyuZhong — CHANGES_REQUESTED** [review]

**Comment by pitrou** [comment]
We have 128-bit integer types in `arrow/util/int128_internal.h`, why not use them always?

**Comment by MingyuZhong** [comment]
> We have 128-bit integer types in `arrow/util/int128_internal.h`, why not use them always?

Because BasicDecimal128 cannot depend on boost per the comment at https://github.com/apache/arrow/blob/3d0a9d58b6fe29dcb208c3fa244c789449517988/cpp/src/arrow/util/basic_decimal.h#L41

**Comment by pitrou** [comment]
Ah, thank you!

File: cpp/src/arrow/util/basic_decimal.cc
Comment (Luminarys) [reply cpp/src/arrow/util/basic_decimal.cc]: Done.

File: cpp/src/arrow/util/basic_decimal.cc
Comment (Luminarys) [reply cpp/src/arrow/util/basic_decimal.cc]: Oops, didn't catch this when making the change to use native types. It also turns out the unit tests didn't catch this either, I've added some more extensive ones for this situation.

File: cpp/src/arrow/util/basic_decimal.cc
Comment (MingyuZhong) [inline cpp/src/arrow/util/basic_decimal.cc:257]: kInt64Mask

**Review by MingyuZhong — CHANGES_REQUESTED** [review]

File: cpp/src/arrow/util/basic_decimal.cc
Comment (Luminarys) [reply cpp/src/arrow/util/basic_decimal.cc:257]: This should be done already.

Comment (Luminarys) [reply cpp/src/arrow/util/basic_decimal.cc:257]: Oh I see, will fix.

**Comment by Luminarys** [comment]
Looks like the tests all use the native branch so the 64 * 64 -> 128 helper is unused. Should we skip the native fallback for the 128 * 128 -> 128 for now and add it in the Decimal256 changes?

**Comment by MingyuZhong** [comment]
> Looks like the tests all use the native branch so the 64 * 64 -> 128 helper is unused. Should we skip the native fallback for the 128 * 128 -> 128 for now and add it in the Decimal256 changes?

There are checks that disable the native branch. For example, https://github.com/apache/arrow/pull/8279/checks?check_run_id=1177973984. It failed because kInt64Mask is unused when the native branch is disabled. Please fix it.

**Comment by Luminarys** [comment]
> > Looks like the tests all use the native branch so the 64 * 64 -> 128 helper is unused. Should we skip the native fallback for the 128 * 128 -> 128 for now and add it in the Decimal256 changes?
> 
> There are checks that disable the native branch. For example, https://github.com/apache/arrow/pull/8279/checks?check_run_id=1177973984. It failed because kInt64Mask is unused when the native branch is disabled. Please fix it.

Done.

**Comment by MingyuZhong** [comment]
LGTM

**Comment by pitrou** [comment]
@Luminarys Can you take a look at the CI failures and try to fix them?

**Comment by emkornfield** [comment]
@Luminarys looking at the first CI failure:

It is:
```
void ExtendAndMultiplyUint64(uint64_t x, uint64_t y, uint64_t* hi, uint64_t* lo) {
```

**Comment by Luminarys** [comment]
Wrapped the method in a guard with a TODO to remove later.

File: cpp/src/arrow/util/basic_decimal.cc
Comment (emkornfield) [inline cpp/src/arrow/util/basic_decimal.cc:259]: docs on what hi and lo would be useful.

File: cpp/src/arrow/util/basic_decimal.cc
Comment (emkornfield) [inline cpp/src/arrow/util/basic_decimal.cc:270]: some documentation on the algorithm might make sense.  maybe a link to an algorithm in: https://en.wikipedia.org/wiki/Multiplication_algorithm

File: cpp/src/arrow/util/basic_decimal.cc
Comment (emkornfield) [inline cpp/src/arrow/util/basic_decimal.cc:297]: can you comment on why hi is hi is being reset here (and lo isn't being used at all?)

**Comment by emkornfield** [comment]
Mostly looks good, one question about resetting hi which I don't quite understand and a couple of areas where documentation could make things clearer.  After these are addressed, I'll merge unless @pitrou has more comments.

File: cpp/src/arrow/util/basic_decimal.cc
Comment (Luminarys) [reply cpp/src/arrow/util/basic_decimal.cc:297]: Added a comment, but note that lo is used (since it's an output parameter in ExtendAndMultiplyUint64)

File: cpp/src/arrow/util/basic_decimal.cc
Comment (Luminarys) [reply cpp/src/arrow/util/basic_decimal.cc:284]: Done

File: cpp/src/arrow/util/basic_decimal.cc
Comment (Luminarys) [reply cpp/src/arrow/util/basic_decimal.cc:261]: Done

**Comment by emkornfield** [comment]
+1 merging.

**Comment by emkornfield** [comment]
@Luminarys do you have a JIRA account?  Would you mind creating one so we can give you credit for the work?

**Comment by Luminarys** [comment]
Thanks, my JIRA handle is eumen, not sure how to assign the issue to myself.

**Comment by emkornfield** [comment]
@Luminarys you need to be added as a contributor to be able get assigned issues.  I just did this and assigned the issue to you

**Pull request closed without merging**
